sidecar/cook/sidecar/progress.py

```diff
@@ -0,0 +1,127 @@
+#  sell copies of the Software, and to permit persons to whom the Software is
+#  furnished to do so, subject to the following conditions:
+#
+#  The above copyright notice and this permission notice shall be included in
+#  all copies or substantial portions of the Software.
+#
+#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+#  IN THE SOFTWARE.
+#
+"""Entrypoint for Cook's sidecar progress reporter."""
+
+import faulthandler
+import logging
+import os
+import requests
+import signal
+import sys
+
+import cook.sidecar.config as csc
+import cook.sidecar.tracker as cst
+from cook.sidecar.version import __version__
+
+
+def start_progress_trackers():
+    try:
+        config = csc.initialize_config(os.environ)
+
+        default_url = config.callback_url
+        current_url = default_url
+
+        def send_progress_message(message):
+            nonlocal current_url
+            try:
+                for i in range(config.max_post_attempts):
+                    response = requests.post(current_url, allow_redirects=False, timeout=config.max_post_time_secs, json=message)
+                    if response.status_code == 202:
+                        return True
+                    elif response.is_redirect and response.status_code == 307:
+                        current_url = response.headers['location']
+                        logging.info(f'Redirected! Changed progress update callback url to: {current_url}')
+                    else:
+                        logging.warning(f'Unexpected progress update response ({response.status_code}): {response.content}')
+                        break
+                else:
+                    logging.warning(f'Reached max redirect retries ({config.max_post_redirect_follow})')
+            except Exception:
+                logging.exception(f'Error raised while posting progress update to {current_url}')
+            current_url = default_url
+            logging.info(f'Failed to post progress update. Reset progress update callback url: {current_url}')
+            return False
+
+        max_message_length = config.max_message_length
+        sample_interval_ms = config.progress_sample_interval_ms
+        sequence_counter = cst.ProgressSequenceCounter()
+        progress_updater = cst.ProgressUpdater(max_message_length, sample_interval_ms, send_progress_message)
+
+        def launch_progress_tracker(progress_location, location_tag):
+            logging.info(f'Location {progress_location} tagged as [tag={location_tag}]')
```

Comment:
Please see https://github.com/twosigma/Cook/pull/1368

Reply:
🤦‍♂
If we're going to continue changing the executor code then maintaining these two separate codebases might become a problem.

Reply:
This was helpful additional logging that I was missing while debugging something recently.

Reply:
✔️